mmv1/third_party/terraform/services/container/resource_container_cluster.go.tmpl

```diff
@@ -1722,13 +1728,120 @@ func ResourceContainerCluster() *schema.Resource {
 				ConflictsWith: []string{"enable_autopilot"},
 			},
 
+			"control_plane_endpoints_config": {
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				Optional:    true,
```

Comment:
```suggestion
				Computed:    true,
				Optional:    true,
```

Reply:
Since this gets returned back when the settings are set through the old mechanisms, this *must* be set.

Reply:
The docs at https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#computed were very unclear as to what Computed actually does. I gathered that it seems to have an effect on computed diffs, but that is not mentioned at all in the docs. I was largely cargo-culting this attribute and applying it to OUTPUT_ONLY fields. Does this mean dns_endpoint_config needs to be Computed as well, since the DNS endpoint field is always present? I will assume yes for now.

Reply:
We talk about the option here in our docs, in case it helps! https://googlecloudplatform.github.io/magic-modules/develop/permadiff/#default. HashiCorp under-covers Optional+Computed, as it is required much more often in GCP which has individually larger resources with more moving parts than AWS, Azure, or most other providers.

We use this if the API will return a value when `null` (i.e. nothing, although explicit nulls do the same) is sent and it's not a known fixed default for a primitive.

> Does this mean dns_endpoint_config needs to be Computed as well, since the DNS endpoint field is always present? I will assume yes for now.

Probably yes!

Reply:
FYI, after going through more of the tests, I found another implication of the use of Computed on the top level fields like `control_plane_endpoints_config` and `ip_endpoints_config`. I had assumed that since the leaf fields were not marked Computed, that I would be following https://developer.hashicorp.com/terraform/plugin/sdkv2/best-practices/deprecations#example-renaming-of-an-optional-attribute. But, apparently TF records the state of all the subfields, and so by only setting one of the fields in the flatten function (corresponding to whichever the user configured), the other unset fields would record their default values. This caused problems with unexpected diffs. So I changed to setting all the fields all the time in flatten. It seems like things are working now, I'm running the tests one last time locally to be sure it all passes.